tests/InferTypesFuzz.elm

```diff
@@ -0,0 +1,329 @@
+    -- Does not produce NaNs, but that should not be an issue for us.
+    Random.float (-1.0 / 0.0) (1.0 / 0.0)
+        |> Random.map (literal Float)
+
+
+boolExpr : Generator Canonical.LocatedExpr
+boolExpr =
+    Random.bool
+        |> Random.map (literal Bool)
+
+
+charExpr : Generator Canonical.LocatedExpr
+charExpr =
+    Random.int 0 0x0010FFFF
+        |> Random.map Char.fromCode
+        |> Random.map (literal Char)
+
+
+stringExpr : Generator Canonical.LocatedExpr
+stringExpr =
+    Random.int 0 0x0010FFFF
+        |> Random.list 10
+        |> Random.map (List.map Char.fromCode)
+        |> Random.map String.fromList
+        |> Random.map (literal String)
+
+
+unitExpr : Generator Canonical.LocatedExpr
+unitExpr =
+    Canonical.Unit
+        |> located
+        |> Random.constant
+
+
+literal : (a -> Literal) -> a -> Canonical.LocatedExpr
+literal wrap value =
+    value
+        |> wrap
+        |> Canonical.Literal
+        |> located
+
+
+listExpr : Type -> Generator Canonical.LocatedExpr
+listExpr elementType =
+    elementType
+        |> exprGenerator
+        |> Random.list 10
+        |> Random.map Canonical.List
+        |> Random.map located
+
+
+intToIntFunctionExpr : Generator Canonical.LocatedExpr
+intToIntFunctionExpr =
+    let
+        combine argument intPart =
+            Canonical.lambda argument <|
+                located <|
+                    Canonical.Plus
+                        (located <| Canonical.Argument argument)
+                        intPart
+
+        intSubExpr =
+            exprGenerator Type.Int
+    in
+    Random.map2 combine
+        -- TODO: Later we will need something better to avoid shadowing.
+        randomVarName
+        intSubExpr
+        |> Random.map located
+
+
+randomVarName : Generator VarName
+randomVarName =
+    let
+        starters =
+            "abcdefghijklnmoqprstuvwxyz"
+
+        others =
+            "ABCDEFGHIJKLMNOQPRSTUVQXYZ_0123456789"
+
+        all =
+            starters ++ others
+
+        charFrom string =
+            string
+                |> String.toList
+                |> List.map Random.constant
+                |> Random.choices (Random.constant 'a')
+
+        firstChar =
+            charFrom starters
+
+        rest =
+            Random.list 5 <| charFrom all
+    in
+    Random.map2 (::) firstChar rest
+        |> Random.map String.fromList
+        |> Random.map VarName
+
+
+dumpType : Type -> String
+dumpType type_ =
+    type_
+        |> Type.toString Type.emptyState
+        |> Tuple.first
+
+
+{-| An expression shrinker that preserves the inferred type.
+-}
+exprShrinker : Shrinker Canonical.LocatedExpr
+exprShrinker expr =
+    case Located.unwrap expr of
+        Canonical.Literal lit ->
+            lit
+                |> shrinkLiteral
+                |> Shrink.map Canonical.Literal
+                |> Shrink.map located
+
+        Canonical.Plus left right ->
+            expr
+                |> shrinkPlus left right
```

Comment:
The Shrinker API is really weird. I'd rather have the right amount of arguments and not give it useless arguments, bearing the cost of not being able to have a type annotation...
`shrinkPlus left right`
```elm
{-| shrinkPlus : Canonical.LocatedExpr -> Canonical.LocatedExpr -> LazyList Canonical.LocatedExpr
-}
shrinkPlus left right =
  lazyMap2 Canonical.Plus
    (exprShrinker left)
    (exprShrinker right)
    |> Shrink.map located
```
Would that work? (I haven't tried to run it)

Reply:
It's nice to see that a new version of the API will hide LazyList totally and allow you to `Simplify.fromFunction : (a -> List a) -> Simplifier a`. Can't wait!
https://github.com/elm-explorations/test/blob/master/src/Simplify.elm